apps/opencs/view/render/terraintexturemode.cpp

```diff
@@ -0,0 +1,328 @@
+{
+    return false;
+}
+
+bool CSVRender::TerrainTextureMode::secondarySelectStartDrag (const QPoint& pos)
+{
+    return false;
+}
+
+void CSVRender::TerrainTextureMode::drag (const QPoint& pos, int diffX, int diffY, double speedFactor) {
+}
+
+void CSVRender::TerrainTextureMode::dragCompleted(const QPoint& pos) {
+}
+
+void CSVRender::TerrainTextureMode::dragAborted() {
+}
+
+void CSVRender::TerrainTextureMode::dragWheel (int diff, double speedFactor) {}
+
+void CSVRender::TerrainTextureMode::handleDragEnterEvent (QDragEnterEvent *event) {
+    event->accept();
+}
+
+void CSVRender::TerrainTextureMode::handleDropEvent (QDropEvent *event) {
+  const CSMWorld::TableMimeData* mime = dynamic_cast<const CSMWorld::TableMimeData*> (event->mimeData());
+
+  if (!mime) // May happen when non-records (e.g. plain text) are dragged and dropped
+      return;
+
+  if (mime->holdsType (CSMWorld::UniversalId::Type_LandTexture))
+  {
+      const std::vector<CSMWorld::UniversalId> ids = mime->getData();
+
+      for (const CSMWorld::UniversalId& uid : ids)
+      {
+          mBrushTexture = uid.getId();
+          emit passBrushTexture(mBrushTexture);
+      }
+  }
+}
+
+void CSVRender::TerrainTextureMode::editTerrainTextureGrid(const WorldspaceHitResult& hit)
+{
+    mCellId = getWorldspaceWidget().getCellId (hit.worldPos);
+    std::string hash = "#";
+    std::string space = " ";
+    std::size_t hashlocation = mCellId.find(hash);
+    std::size_t spacelocation = mCellId.find(space);
+    std::string slicedX = mCellId.substr (hashlocation+1, spacelocation-hashlocation);
+    std::string slicedY = mCellId.substr (spacelocation+1);
+    CSMWorld::CellCoordinates mCoordinates(stoi(slicedX), stoi(slicedY));
+
+    CSMDoc::Document& document = getWorldspaceWidget().getDocument();
+    CSMWorld::IdTable& landTable = dynamic_cast<CSMWorld::IdTable&> (
+        *document.getData().getTableModel (CSMWorld::UniversalId::Type_Land));
+    CSMWorld::IdTable& ltexTable = dynamic_cast<CSMWorld::IdTable&> (
+        *document.getData().getTableModel (CSMWorld::UniversalId::Type_LandTextures));
+
+    int textureColumn = landTable.findColumnIndex(CSMWorld::Columns::ColumnId_LandTexturesIndex);
+    CSMWorld::LandTexturesColumn::DataType mPointer = landTable.data(landTable.getModelIndex(mCellId, textureColumn)).value<CSMWorld::LandTexturesColumn::DataType>();
+    CSMWorld::LandTexturesColumn::DataType mNew(mPointer);
+
+    int xInCell ((hit.worldPos.x() - (stoi(slicedX)* cellSize)) * landTextureSize / cellSize);
+    int yInCell ((hit.worldPos.y() - (stoi(slicedY)* cellSize)) * landTextureSize / cellSize);
+
+    hashlocation = mBrushTexture.find(hash);
+    std::string mBrushTextureInt = mBrushTexture.substr (hashlocation+1);
+    int brushInt = stoi(mBrushTexture.substr (hashlocation+1))+1; // All indices are offset by +1
+
+    if (mBrushShape == 0) mNew[yInCell*landTextureSize+xInCell] = brushInt;
```

Comment:
A better way to handle this would be via Brush classes (with a shared brush base class).

Reply:
A new class that would have functions like getShape, getSize, setSize, setShape etc.? That would probably affect many parts of the code, and I regarding other editmodes, I wonder if this base class inheriting would be something like BrushBase, TextureEditBrush : BrushBase, LandShapeEditBrush : BrushBase etc.? I guess size and shape are only common things needed for all brushes. Another way to handle this would be to have all brush settings mashed into one class, and therefore have lots of unnecessary variables, but that's probably not an optimal solution.

Handling a Brush class like that also makes me a bit cautious (slower to code) as there's more to think of regarding RAII and dynamic memory allocation. I've tried to stay on RAII-side and use automated memory handling as much as possible, but I wonder if this is a case where custom destructors and copy constructors will be needed.